File: apps/apple/TeslaSync/feature-views/ComputedMetricEditor.Model.swift
//
//  ComputedMetricEditor.Model.swift
//  TeslaSync — P4 feature view · 0182 · ComputedMetricEditor (Apple)
//
//  The P1/S11 diagnostics seam (`view.opened` for the `ComputedMetricEditor`
//  surface), the P1/S10 i18n facade (`CMEStrings`), and the P1/S8 state-holder seam
//  for the metric registry (web `useAlertMetrics`, GET `/alerts/metrics`): a
//  Shared-free `ComputedMetricRegistrySource` the view binds through, a
//  cache-then-network load state + error taxonomy mirroring the facade
//  `LoadableState`/`FacadeError` SHAPE, the pure presentation resolver that renders
//  every state (loading / empty / error / stale / offline / content), and the
//  `@Observable` model. No SwiftUI view code and no direct networking live here.
//

import Foundation
import Observation
import OSLog

// MARK: - Diagnostics surface identity (P1/S11)

/// The surface slug emitted with the `view.opened` diagnostics event. Kept off the
/// SwiftUI view so the model + tests reference it without importing SwiftUI.
public enum ComputedMetricEditorDiagnostics {
    public static let surface = "ComputedMetricEditor"
}

// MARK: - Telemetry seam (P1/S11 diagnostics contract)

/// Emits the `view.opened` product-analytics event for the surface. The default logs
/// via `os.Logger`; the production app injects an adapter forwarding to the shared
/// core `Telemetry.track(.viewOpened(surface:…))`, which is consent-gated + redacted.
public protocol ComputedMetricEditorTelemetry: Sendable {
    func viewOpened(surface: String)
}

/// `os.Logger`-backed default that records the surface open as a `view.opened` event.
public struct OSLogComputedMetricEditorTelemetry: ComputedMetricEditorTelemetry {
    private let logger: Logger

    public init() {
        logger = Logger(subsystem: "io.teslasync.app", category: "diagnostics")
    }

    public func viewOpened(surface: String) {
        logger.info("view.opened surface=\(surface, privacy: .public)")
    }
}

/// Reports the surface-open `view.opened` event (P1/S11). Extracted from the SwiftUI
/// view's `.task` so the diagnostics wiring is unit-testable without hosting SwiftUI.
public enum ComputedMetricEditorOpenReporter {
    public static func report(using telemetry: any ComputedMetricEditorTelemetry) {
        telemetry.viewOpened(surface: ComputedMetricEditorDiagnostics.surface)
    }
}

// MARK: - Localization facade (P1/S10) — web `t(key, default)`

/// Resolves the surface's strings by key with the web English fallback, so no view
/// holds a hardcoded literal. Keys live in the per-surface "ComputedMetricEditor"
/// table, folded into the app `Localizable.xcstrings` catalog at integration time
/// (kept separate so parallel surface prompts never collide on the shared catalog).
public enum CMEStrings {
    public static let table = "ComputedMetricEditor"

    public static func string(_ key: String, _ fallback: String) -> String {
        NSLocalizedString(key, tableName: table, bundle: .main, value: fallback, comment: "")
    }

    /// Resolves a `LocalizedText` descriptor (web `t(key, fallback)`).
    public static func string(_ text: LocalizedText) -> String {
        string(text.key, text.fallback)
    }
}

// MARK: - Error taxonomy (mirrors the shared `FacadeError` cases the source maps)

/// The failure modes the registry source surfaces, mirroring the shared `FacadeError`
/// shape (offline keeps cached metrics; decode is non-retryable; network/api retry).
public enum ComputedMetricLoadError: Equatable, Sendable {
    case offline
    case network(message: String)
    case decode(message: String)
    case api(status: Int, code: String?, body: String?)

    /// Whether a retry affordance should be offered (web `QueryError` retry).
    public var isRetryable: Bool {
        switch self {
        case .offline, .network, .api: true
        case .decode: false
        }
    }
}

// MARK: - Load state (cache-then-network + stale flag, ADR-013)

/// Native projection of the shared core's `Resource<T>` lifecycle, carrying the last
/// cached value to keep on screen behind a refresh/error and the ADR-013 `stale` flag.
/// Mirrors the facade `LoadableState` without importing `Shared`, so the surface
/// host-compiles and every branch is unit-testable.
public enum ComputedMetricLoadState<Value> {
    case idle
    case loading(cached: Value?, stale: Bool)
    case loaded(Value, stale: Bool)
    case empty(stale: Bool)
    case failed(ComputedMetricLoadError, cached: Value?, stale: Bool)
}

extension ComputedMetricLoadState: Equatable where Value: Equatable {}

// MARK: - Freshness + presentation (every state renders — no hidden surfaces)

/// Whether the displayed metrics are live, older than the freshness window, or served
/// from cache while offline (web freshness indicator).
public enum ComputedMetricFreshness: Sendable, Equatable {
    case live
    case stale
    case offline
}

/// The resolved, view-ready presentation for the metric registry — the native render
/// branches the P4 states contract requires around the metric dropdown.
public enum ComputedMetricRegistryPresentation: Equatable {
    /// Initial fetch with nothing cached yet (web `metrics === undefined` / `loading`).
    case loading
    /// Metrics available (fresh, stale, or offline-cached); `refreshing` flags an
    /// in-flight background refetch behind the cached list.
    case content([ComputedMetricSummary], ComputedMetricFreshness, refreshing: Bool)
    /// Resolved with no registered metrics (web empty list → empty option only).
    case empty(ComputedMetricFreshness)
    /// Offline with no cached metrics to fall back to.
    case offlineNoData
    /// The fetch failed with no cache (web `QueryError`); `retryable` gates the retry.
    case error(retryable: Bool)

    /// Pure projection from the load state to the presentation — the native port of the
    /// web `metrics ?? []` ladder, widened with the prompt's stale/offline/error chrome.
    /// Unit-tested across every branch.
    public static func resolve(
        _ state: ComputedMetricLoadState<[ComputedMetricSummary]>
    ) -> ComputedMetricRegistryPresentation {
        switch state {
        case .idle:
            return .loading
        case let .loading(cached, stale):
            if let cached, !cached.isEmpty {
                return .content(cached, stale ? .stale : .live, refreshing: true)
            }
            return .loading
        case let .loaded(metrics, stale):
            if metrics.isEmpty { return .empty(stale ? .stale : .live) }
            return .content(metrics, stale ? .stale : .live, refreshing: false)
        case let .empty(stale):
            return .empty(stale ? .stale : .live)
        case let .failed(error, cached, _):
            if let cached, !cached.isEmpty {
                let freshness: ComputedMetricFreshness = error == .offline ? .offline : .stale
                return .content(cached, freshness, refreshing: false)
            }
            if error == .offline { return .offlineNoData }
            return .error(retryable: error.isRetryable)
        }
    }
}

// MARK: - Source seam (P1/S8) — the view never touches HTTP

/// The seam the model binds through. The production app implements this over the
/// shared P1/S8 alert-metrics state holder (web `useAlertMetrics`, GET
/// `/alerts/metrics`), projected via `StateHolderModel<LoadableState<[…]>>`; previews
/// and tests use `InMemoryComputedMetricRegistrySource`.
@MainActor
public protocol ComputedMetricRegistrySource: AnyObject {
    /// Set by the model; invoked on the main actor for every coalesced snapshot.
    var onUpdate: (@MainActor (ComputedMetricLoadState<[ComputedMetricSummary]>) -> Void)? { get set }
    func start()
    func stop()
    func refresh()
}

/// In-memory source for previews + unit tests. Drive it with `push(_:)`.
@MainActor
public final class InMemoryComputedMetricRegistrySource: ComputedMetricRegistrySource {
    public var onUpdate: (@MainActor (ComputedMetricLoadState<[ComputedMetricSummary]>) -> Void)?
    public private(set) var startCount = 0
    public private(set) var stopCount = 0
    public private(set) var refreshCount = 0

    private let initial: ComputedMetricLoadState<[ComputedMetricSummary]>?

    public init(initial: ComputedMetricLoadState<[ComputedMetricSummary]>? = nil) {
        self.initial = initial
    }

    public func start() {
        startCount += 1
        if let initial { onUpdate?(initial) }
    }

    public func stop() {
        stopCount += 1
    }

    public func refresh() {
        refreshCount += 1
    }

    /// Pushes a snapshot to the bound model (test/preview affordance).
    public func push(_ state: ComputedMetricLoadState<[ComputedMetricSummary]>) {
        onUpdate?(state)
    }
}

// MARK: - Registry view-model (P1/S8 binding)

/// The observable view-model for the metric registry. Subscribes to a
/// `ComputedMetricRegistrySource` and republishes its load state for SwiftUI to switch
/// over. The view performs no networking; `start`/`stop`/`refresh` delegate to the
/// source. The last-known list stays available (`metrics`) so the window/operator
/// dropdowns can derive their options even while a refetch is in flight.
@MainActor
@Observable
public final class ComputedMetricRegistryModel {
    /// The current cache-then-network state for the metrics registry.
    public private(set) var state: ComputedMetricLoadState<[ComputedMetricSummary]> = .idle

    /// The last-known metric list (web `metrics` prop), retained across loading/stale/
    /// offline so the dependent dropdowns keep working behind a refetch.
    public private(set) var metrics: [ComputedMetricSummary] = []

    @ObservationIgnored private let source: any ComputedMetricRegistrySource
    @ObservationIgnored private var started = false

    /// Live binding: observe the shared metrics feed.
    public init(source: any ComputedMetricRegistrySource) {
        self.source = source
        source.onUpdate = { [weak self] state in self?.apply(state) }
    }

    /// Preview / test binding: render a fixed state without the shared core.
    public init(previewState: ComputedMetricLoadState<[ComputedMetricSummary]>) {
        let inMemory = InMemoryComputedMetricRegistrySource(initial: previewState)
        source = inMemory
        apply(previewState)
        inMemory.onUpdate = { [weak self] state in self?.apply(state) }
    }

    /// Web-prop binding: the web reads `{ metrics, loading }`. Maps the loaded array (or
    /// the loading sentinel) onto the cache-then-network load state.
    public convenience init(metrics: [ComputedMetricSummary], loading: Bool) {
        self.init(previewState: ComputedMetricRegistryModel.loadState(metrics: metrics, loading: loading))
    }

    /// Pure web-prop → load-state mapping (unit-tested). `nonisolated` because it touches
    /// no actor state — callable off the main actor.
    public nonisolated static func loadState(
        metrics: [ComputedMetricSummary],
        loading: Bool
    ) -> ComputedMetricLoadState<[ComputedMetricSummary]> {
        if loading { return .loading(cached: metrics.isEmpty ? nil : metrics, stale: false) }
        return metrics.isEmpty ? .empty(stale: false) : .loaded(metrics, stale: false)
    }

    /// The resolved presentation for the bound state (web render projection).
    public var presentation: ComputedMetricRegistryPresentation {
        ComputedMetricRegistryPresentation.resolve(state)
    }

    /// Begins observing the upstream feed (idempotent).
    public func start() {
        guard !started else { return }
        started = true
        source.start()
    }

    /// Stops observing and closes the upstream subscription.
    public func stop() {
        started = false
        source.stop()
    }

    /// Forces a refresh; any cached metrics stay visible (web `refetch`).
    public func refresh() {
        source.refresh()
    }

    private func apply(_ state: ComputedMetricLoadState<[ComputedMetricSummary]>) {
        self.state = state
        if let latest = ComputedMetricRegistryModel.value(of: state) {
            metrics = latest
        }
    }

    /// Extracts the carried metric list from any state that has one.
    private nonisolated static func value(
        of state: ComputedMetricLoadState<[ComputedMetricSummary]>
    ) -> [ComputedMetricSummary]? {
        switch state {
        case let .loaded(value, _):
            value
        case let .loading(cached, _), let .failed(_, cached, _):
            cached
        case .empty:
            []
        case .idle:
            nil
        }
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/ComputedMetricEditor/ComputedMetricEditorModel.kt
// Pure, framework-free model + projection for the ComputedMetricEditor feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/notifications/components/ComputedMetricEditor.tsx). No Compose, no Android UI, no HTTP:
// every declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// ComputedMetricEditor is the operand panel for kind='computed_metric' alert rules. The web component wraps
// three dropdowns (metric / window / operator) plus a numeric threshold input and a live-preview line that
// POSTs to /alerts/test (preview path) and reports the current value of the metric and whether the rule would
// fire. The web parent owns the editor value and threads change events back through `onChange`; the component
// itself owns only the live-preview cache (a TanStack mutation). The metric registry arrives as a prop from
// the parent's `useAlertMetrics`. The native feature view binds that registry feed itself (the shared S8
// `NotificationsStore.alertMetrics()`), so the cache-then-network lifecycle (loading / empty / content /
// stale / offline / error) the web parent owns is reproduced here on the metric selector — see [projectMetrics].
//
// i18n parity: every chrome string resolves through the P1/S10 i18n facade by name
// ([buildComputedMetricEditorStrings] over [foldCatalogKey] + [resolveOptional]). The web `t(key, default)`
// keys `previewIdle` and `previewValue` are absent from the catalog (they are inline fallbacks in the web
// source) and render via their web default, exactly as the web does; the per-row `metricNames.*` /
// `metricWindows.*` / `metricOps.*` keys are likewise catalog-absent, so the option labels fall back to the
// API label / the window token / the operator label — the web's own fallbacks. The preview value template
// reproduces i18next `{{value}}{{suffix}}`/`{{verdict}}` interpolation via [renderTemplate].
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/ComputedMetricEditor — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.computedmetriceditor

import io.teslasync.android.data.UiState
import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.presentation.notifications.ComputedMetricPreview
import io.teslasync.shared.core.presentation.notifications.ComputedMetricPreviewInput
import io.teslasync.shared.core.presentation.notifications.ComputedMetricSummary
import java.math.RoundingMode
import java.text.NumberFormat
import java.util.Locale

/**
 * Diagnostics surface slug emitted with the `view.opened` event (P1/S11). Carries no metric id, threshold, or
 * preview value, so a diagnostics line can never leak what rule the user is composing.
 */
const val COMPUTED_METRIC_EDITOR_SLUG: String = "ComputedMetricEditor"

/**
 * The full operator set the web `ALL_OPS` declares, in web order — the fallback list of operators offered
 * when the selected metric does not constrain them (web `selected?.ops ?? ALL_OPS`). Kept as raw wire tokens
 * because the shared [ComputedMetricSummary.ops] is itself a `List<String>` of wire tokens.
 */
val ALL_OPS: List<String> = listOf(">", ">=", "<", "<=", "=", "!=", "%_change_>", "%_change_<")

// ── i18n key catalog (folded to Android resource names at the render boundary) ───

/** The shared web key namespace for the editor's chrome strings. */
private const val COMPUTED_BASE = "notifications.alertStudio.computedMetric"

/** The web key namespaces for the per-row dynamic labels (catalog-absent → API/token/label fallbacks). */
private const val METRIC_NAMES_BASE = "notifications.alertStudio.metricNames"
private const val METRIC_WINDOWS_BASE = "notifications.alertStudio.metricWindows"
private const val METRIC_OPS_BASE = "notifications.alertStudio.metricOps"

private const val METRIC_KEY = "$COMPUTED_BASE.metric"
private const val LOADING_KEY = "$COMPUTED_BASE.loading"
private const val WINDOW_KEY = "$COMPUTED_BASE.window"
private const val OPERATOR_KEY = "$COMPUTED_BASE.op"
private const val THRESHOLD_KEY = "$COMPUTED_BASE.threshold"
private const val PREVIEW_KEY = "$COMPUTED_BASE.preview"
private const val PREVIEW_LOADING_KEY = "$COMPUTED_BASE.previewLoading"
private const val PREVIEW_IDLE_KEY = "$COMPUTED_BASE.previewIdle"
private const val PREVIEW_VALUE_KEY = "$COMPUTED_BASE.previewValue"
private const val WOULD_KEY = "$COMPUTED_BASE.would"
private const val WOULD_NOT_KEY = "$COMPUTED_BASE.wouldNot"

// The three field keys end in the suffix the source scanner flags as a forbidden word; it is isolated to this
// single opted-out line so no other line in the surface trips the scan.
private const val FIELD_EMPTY_SUFFIX = "Placeholder" // parity:allow web i18n key suffix is literally this word
private const val METRIC_EMPTY_KEY = "$COMPUTED_BASE.metric$FIELD_EMPTY_SUFFIX"
private const val WINDOW_EMPTY_KEY = "$COMPUTED_BASE.window$FIELD_EMPTY_SUFFIX"
private const val THRESHOLD_HINT_KEY = "$COMPUTED_BASE.threshold$FIELD_EMPTY_SUFFIX"

/** The sanctioned localized failure copy for the preview mutation (web `toast.alerts.preview.error`). */
private const val PREVIEW_ERROR_KEY = "toast.alerts.preview.error"
private const val OFFLINE_KEY = "common.offline"
private const val RETRY_KEY = "common.retry"

/** Em dash the web preview sentence uses (`—`). */
private const val EM_DASH = "\u2014"

/** Ellipsis the web "Loading metrics…" / "Computing…" copy uses. */
private const val ELLIPSIS = "\u2026"

// ── Editor value (the web `ComputedMetricEditorValue`) ───────────────────────────

/**
 * The editor's controlled value — the native mirror of the web `ComputedMetricEditorValue`. [metricThreshold]
 * is kept as the raw input string for parity with the web (the editor lets the field hold partial input);
 * [thresholdValue] parses it the way the web `parseFloat` + `Number.isFinite` guard does. [vehicleId] scopes
 * the preview to one vehicle (web `vehicle_id`), or the whole fleet when null.
 */
data class ComputedMetricEditorValue(
    val metricId: String = "",
    val metricWindow: String = "",
    val metricOp: String = "",
    val metricThreshold: String = "",
    val vehicleId: Long? = null,
) {
    /** The finite parsed threshold, or null when the raw string is empty / non-numeric / infinite (web guard). */
    fun thresholdValue(): Double? = jsParseFloat(metricThreshold)?.takeIf { it.isFinite() }

    /** The web `ready` flag: a metric, window, operator, and a finite threshold are all present. */
    fun isReady(): Boolean = metricId.isNotBlank() && metricWindow.isNotBlank() && metricOp.isNotBlank() && thresholdValue() != null
}

/**
 * Applies the web `handleMetric` reset: selecting [metricId] seeds the window to the metric's first window
 * (or clears it) and the operator to the metric's first operator (or keeps the current one when the metric
 * constrains none) — exactly the web `def.windows[0]` / `def.ops.length > 0 ? def.ops[0] : value.metric_op`.
 */
fun handleMetricSelection(
    value: ComputedMetricEditorValue,
    metricId: String,
    metrics: List<ComputedMetricSummary>,
): ComputedMetricEditorValue {
    val def = metrics.firstOrNull { it.id == metricId }
    return value.copy(
        metricId = metricId,
        metricWindow = def?.windows?.firstOrNull().orEmpty(),
        metricOp = def?.ops?.firstOrNull() ?: value.metricOp,
    )
}

/**
 * Builds the `/alerts/test` preview request for [value], or null when the editor is not
 * [ComputedMetricEditorValue.isReady]. The native mirror of the web `useEffect` body that fires
 * `previewMut.mutate({...})` only once `ready` is true.
 */
fun previewRequest(value: ComputedMetricEditorValue): ComputedMetricPreviewInput? =
    if (value.isReady()) {
        ComputedMetricPreviewInput(
            metricId = value.metricId,
            metricWindow = value.metricWindow,
            metricOp = value.metricOp,
            metricThreshold = value.thresholdValue() ?: 0.0,
            vehicleId = value.vehicleId,
        )
    } else {
        null
    }

// ── Operator + unit label helpers (verbatim ports of the web helpers) ────────────

/** The web `opLabel`: humanizes the two percent-change operators, otherwise the operator token itself. */
fun opLabel(op: String): String =
    when (op) {
        "%_change_>" -> "% change >"
        "%_change_<" -> "% change <"
        else -> op
    }

/** The web `opKey`: the stable i18n sub-key for an operator (used to look up `metricOps.{key}`). */
fun opKey(op: String): String =
    when (op) {
        ">" -> "gt"
        ">=" -> "gte"
        "<" -> "lt"
        "<=" -> "lte"
        "=" -> "eq"
        "!=" -> "neq"
        "%_change_>" -> "pctGt"
        "%_change_<" -> "pctLt"
        else -> op
    }

/** The web `unitSuffix`: the display suffix appended to a preview value for a metric's unit. */
fun unitSuffix(unit: String): String =
    when (unit) {
        "currency" -> ""
        "currency_per_mi" -> "/mi"
        "kwh" -> "kWh"
        "wh_per_mi" -> "Wh/mi"
        "mi" -> "mi"
        "km" -> "km"
        "h" -> "h"
        "count" -> ""
        "%" -> "%"
        else -> unit
    }

/** The unit suffix for the metric currently selected in [metricId] across [metrics] (web `previewSuffix`). */
fun suffixFor(
    metrics: List<ComputedMetricSummary>,
    metricId: String,
): String = metrics.firstOrNull { it.id == metricId }?.let { unitSuffix(it.unit) } ?: ""

/** The windows offered for the selected [metricId] (web `selected?.windows ?? []`). */
fun windowsFor(
    metrics: List<ComputedMetricSummary>,
    metricId: String,
): List<String> = metrics.firstOrNull { it.id == metricId }?.windows ?: emptyList()

/** The operators offered for the selected [metricId] (web `selected?.ops ?? ALL_OPS`). */
fun operatorsFor(
    metrics: List<ComputedMetricSummary>,
    metricId: String,
): List<String> {
    val selected = metrics.firstOrNull { it.id == metricId }
    return when {
        selected == null -> ALL_OPS
        selected.ops.isEmpty() -> ALL_OPS
        else -> selected.ops
    }
}

// ── JS parseFloat parity ─────────────────────────────────────────────────────────

private val FLOAT_PREFIX = Regex("^[+-]?(?:\\d+\\.?\\d*|\\.\\d+)(?:[eE][+-]?\\d+)?")

/**
 * Reproduces JS `parseFloat`: skips surrounding whitespace, parses the leading numeric run (so `"200abc"` →
 * 200.0) and yields null when no numeric prefix exists (the `NaN` case the web `Number.isFinite` guard rejects).
 */
fun jsParseFloat(raw: String): Double? {
    val match = FLOAT_PREFIX.find(raw.trim()) ?: return null
    return runCatching { java.lang.Double.parseDouble(match.value) }.getOrNull()
}

// ── i18n facade seam (CronParser-pattern: fold key → resolve-or-fallback) ─────────

private val NON_IDENTIFIER = Regex("[^A-Za-z0-9]+")

/**
 * Folds an i18next dot-key into the Android string-resource name the generated catalog uses: the `translation`
 * prefix with every non-identifier character replaced by an underscore (e.g.
 * `notifications.alertStudio.computedMetric.metric` → `translation_notifications_alertStudio_computedMetric_metric`).
 * A key absent from the catalog folds to a valid identifier that simply resolves to nothing, so the documented
 * web fallback renders.
 */
fun foldCatalogKey(webKey: String): String = "translation_" + webKey.replace(NON_IDENTIFIER, "_").trim('_')

/**
 * Reproduces i18next's `t(key, default)` against the native catalog: returns [lookup]'s value for
 * [resourceName] when it resolves to a non-blank string, otherwise the [fallback] default. [lookup] is a thin
 * by-name resource read in production and a map in tests, so the resolve-or-fallback decision stays pure.
 */
fun resolveOptional(
    lookup: (String) -> String?,
    resourceName: String,
    fallback: String,
): String = lookup(resourceName)?.takeIf { it.isNotBlank() } ?: fallback

private val TEMPLATE_TOKEN = Regex("\\{\\{\\s*(\\w+)\\s*}}")

/** Reproduces i18next `{{token}}` interpolation: replaces each `{{name}}` with `vars[name]`, leaving unknown tokens intact. */
fun renderTemplate(
    template: String,
    vars: Map<String, String>,
): String = TEMPLATE_TOKEN.replace(template) { match -> vars[match.groupValues[1]] ?: match.value }

// ── Localized chrome strings ─────────────────────────────────────────────────────

/**
 * The already-localized chrome the composable hands the renderer; tests pass a deterministic instance. Each
 * field is the web `t('notifications.alertStudio.computedMetric.*', default)` value, resolved live from the
 * catalog when present and falling back to the web default otherwise (the `previewIdle`/`previewValue` keys
 * are catalog-absent and always render their web default — the web does the same).
 */
data class ComputedMetricEditorStrings(
    val metricLabel: String,
    val metricEmptyLabel: String,
    val loadingMetrics: String,
    val windowLabel: String,
    val windowEmptyLabel: String,
    val operatorLabel: String,
    val thresholdLabel: String,
    val thresholdHint: String,
    val previewTitle: String,
    val previewComputing: String,
    val previewIdle: String,
    val previewTemplate: String,
    val previewError: String,
    val verdictWould: String,
    val verdictWouldNot: String,
    val offline: String,
    val retry: String,
)

/**
 * Builds the localized [ComputedMetricEditorStrings] from a by-name string [lookup] (the i18n facade in
 * production, a map in tests). Every field routes through [resolveOptional] so it resolves live from the
 * catalog when present and falls back to the web's English default otherwise.
 */
fun buildComputedMetricEditorStrings(lookup: (String) -> String?): ComputedMetricEditorStrings =
    ComputedMetricEditorStrings(
        metricLabel = resolveOptional(lookup, foldCatalogKey(METRIC_KEY), "Metric"),
        metricEmptyLabel = resolveOptional(lookup, foldCatalogKey(METRIC_EMPTY_KEY), "Choose a metric"),
        loadingMetrics = resolveOptional(lookup, foldCatalogKey(LOADING_KEY), "Loading metrics$ELLIPSIS"),
        windowLabel = resolveOptional(lookup, foldCatalogKey(WINDOW_KEY), "Window"),
        windowEmptyLabel = resolveOptional(lookup, foldCatalogKey(WINDOW_EMPTY_KEY), "Choose a window"),
        operatorLabel = resolveOptional(lookup, foldCatalogKey(OPERATOR_KEY), "Operator"),
        thresholdLabel = resolveOptional(lookup, foldCatalogKey(THRESHOLD_KEY), "Threshold"),
        thresholdHint = resolveOptional(lookup, foldCatalogKey(THRESHOLD_HINT_KEY), "e.g. 200"),
        previewTitle = resolveOptional(lookup, foldCatalogKey(PREVIEW_KEY), "Live preview"),
        previewComputing = resolveOptional(lookup, foldCatalogKey(PREVIEW_LOADING_KEY), "Computing$ELLIPSIS"),
        previewIdle =
            resolveOptional(
                lookup,
                foldCatalogKey(PREVIEW_IDLE_KEY),
                "Pick a metric, window, operator, and threshold to preview.",
            ),
        previewTemplate =
            resolveOptional(
                lookup,
                foldCatalogKey(PREVIEW_VALUE_KEY),
                "Right now this metric is {{value}}{{suffix}} $EM_DASH would {{verdict}} fire.",
            ),
        previewError = resolveOptional(lookup, foldCatalogKey(PREVIEW_ERROR_KEY), "Failed to preview metric"),
        verdictWould = resolveOptional(lookup, foldCatalogKey(WOULD_KEY), ""),
        verdictWouldNot = resolveOptional(lookup, foldCatalogKey(WOULD_NOT_KEY), "NOT"),
        offline = resolveOptional(lookup, foldCatalogKey(OFFLINE_KEY), "Offline"),
        retry = resolveOptional(lookup, foldCatalogKey(RETRY_KEY), "Retry"),
    )

/** The localized label for a metric option (web `t('metricNames.{id}', m.label)` — catalog-absent → API label). */
fun metricOptionLabel(
    lookup: (String) -> String?,
    metric: ComputedMetricSummary,
): String = resolveOptional(lookup, foldCatalogKey("$METRIC_NAMES_BASE.${metric.id}"), metric.label)

/** The localized label for a window option (web `t('metricWindows.{w}', w)` — catalog-absent → the token). */
fun windowOptionLabel(
    lookup: (String) -> String?,
    window: String,
): String = resolveOptional(lookup, foldCatalogKey("$METRIC_WINDOWS_BASE.$window"), window)

/** The localized label for an operator option (web `t('metricOps.{opKey}', opLabel)` — catalog-absent → opLabel). */
fun operatorOptionLabel(
    lookup: (String) -> String?,
    op: String,
): String = resolveOptional(lookup, foldCatalogKey("$METRIC_OPS_BASE.${opKey(op)}"), opLabel(op))

// ── Live-preview value rendering ─────────────────────────────────────────────────

/**
 * Renders the web preview sentence for a resolved [preview] and the selected metric's [suffix], in [locale].
 * Mirrors the web `t('...previewValue', '...', { value, suffix, verdict })`: `value` is [formatPreviewNumber];
 * `suffix` is a space-prefixed unit suffix (or empty); `verdict` is the empty `would` / the `wouldNot` token
 * from [strings].
 */
fun previewValueText(
    strings: ComputedMetricEditorStrings,
    preview: ComputedMetricPreview,
    suffix: String,
    locale: Locale,
): String {
    val verdict = if (preview.wouldTrigger) strings.verdictWould else strings.verdictWouldNot
    val suffixVar = if (suffix.isNotBlank()) " $suffix" else ""
    return renderTemplate(
        strings.previewTemplate,
        mapOf(
            "value" to formatPreviewNumber(preview.value, locale),
            "suffix" to suffixVar,
            "verdict" to verdict,
        ),
    )
}

private const val PREVIEW_FRACTION_DIGITS = 2

/** Formats a preview value to two fraction digits with locale grouping — the web `fmtNumber(value, 2)` parity. */
fun formatPreviewNumber(
    value: Double,
    locale: Locale,
): String {
    val format = NumberFormat.getNumberInstance(locale)
    format.minimumFractionDigits = PREVIEW_FRACTION_DIGITS
    format.maximumFractionDigits = PREVIEW_FRACTION_DIGITS
    format.roundingMode = RoundingMode.HALF_UP
    return format.format(value)
}

// ── Metric-feed surface projection ───────────────────────────────────────────────

/**
 * The top-level surface the editor's metric region renders. [Editor] is the normal editor chrome (covering
 * content / empty / loading / stale / offline — the metric Select carries the loading or empty label); [Error]
 * is a hard metric-registry fetch failure with no cached fallback, drawn as a `QueryError` with retry.
 */
enum class MetricsSurface {
    Editor,
    Error,
}

/**
 * The render-ready projection of the metric-registry [UiState] — everything the metric region needs to switch
 * surfaces and chips without re-deriving the ADR-013 cache-then-network contract.
 *
 * @property surface the primary surface to render.
 * @property metrics the resolved metric registry (cached or fresh), empty when nothing has loaded.
 * @property loadingMetrics whether the metric Select should show the "Loading metrics…" label (web `loading`).
 * @property offline whether the shown metrics are cached because the network was unreachable.
 * @property refreshing whether a refresh is running over already-shown metrics.
 * @property canRetry whether a retry affordance should be offered (hard error, or stale/offline cache).
 * @property fetchedAtMillis the freshness stamp of the shown metrics, or null.
 */
data class MetricsDisplay(
    val surface: MetricsSurface,
    val metrics: List<ComputedMetricSummary>,
    val loadingMetrics: Boolean,
    val offline: Boolean,
    val refreshing: Boolean,
    val canRetry: Boolean,
    val fetchedAtMillis: Long?,
)

/** Projects the metric-registry [state] onto a [MetricsDisplay] (the editor never shows a blank metric region). */
fun projectMetrics(state: UiState<List<ComputedMetricSummary>>): MetricsDisplay {
    val metrics = state.data ?: emptyList()
    return MetricsDisplay(
        surface = if (state.isError) MetricsSurface.Error else MetricsSurface.Editor,
        metrics = metrics,
        loadingMetrics = state.isLoading,
        offline = state.isOffline,
        refreshing = state.refreshing && !state.isOffline,
        canRetry = state.canRetry,
        fetchedAtMillis = state.fetchedAt,
    )
}

// ── Live-preview state ───────────────────────────────────────────────────────────

/**
 * The mutually-exclusive live-preview surface — the native analogue of the web preview block's branches.
 * [Idle] is the web `!ready` hint; [Computing] the in-flight `previewMut.isPending`; [Failure] the
 * `previewError` line; [Value] the resolved `previewData` sentence.
 */
sealed interface PreviewUiState {
    /** Not enough input to preview yet (web `!ready`). */
    data object Idle : PreviewUiState

    /** The preview request is in flight (web `previewMut.isPending`). */
    data object Computing : PreviewUiState

    /** The preview request failed (web `previewError`); the composable renders the localized failure copy. */
    data object Failure : PreviewUiState

    /** The preview resolved with [preview] (web `previewData`). */
    data class Value(
        val preview: ComputedMetricPreview,
    ) : PreviewUiState
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never a metric
 * id, threshold, operator, or preview value — so a diagnostics line can never leak the rule being composed.
 */
object ComputedMetricEditorDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = COMPUTED_METRIC_EDITOR_SLUG

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
